pkg/internal/client/txn.go

```diff
@@ -572,9 +572,38 @@ func (txn *Txn) AddCommitTrigger(trigger func()) {
 	txn.commitTriggers = append(txn.commitTriggers, trigger)
 }
 
+// maybeFinishReadonly provides a fast-path for finishing a read-only
+// transaction without going through the overhead of creating an
+// EndTransactionRequest only to not send it.
+//
+// NB: The logic here must be kept in sync with the logic in txn.Send.
+func (txn *Txn) maybeFinishReadonly(commit bool, deadline *hlc.Timestamp) (bool, *roachpb.Error) {
+	txn.mu.Lock()
+	defer txn.mu.Unlock()
+	if txn.mu.Proto.Writing || txn.mu.writingTxnRecord {
+		return false, nil
+	}
+	txn.mu.finalized = true
+	// Check that read only transactions do not violate their deadline. This can NOT
+	// happen since the txn deadline is normally updated when it is about to expire
```

Comment:
What happens in this case:

1. our txn starts at ts1
1. catches uncertainty read error
1. updates its timestamp
1. new timestamp violates deadline
1. txn retries the read
1. commit fails - only thanks to this code path?